sync15-adapter/src/state.rs

```diff
@@ -46,3 +73,568 @@ impl GlobalState {
+            match &self.info_configuration {
+                Ok(config) => Ok(config.clone()),
+                Err(_) => Err(ErrorKind::StorageHttpError {
+                    code: reqwest::StatusCode::InternalServerError,
+                    route: "info/configuration".to_string(),
+                }.into()),
+            }
+        }
+
+        fn fetch_info_collections(&self) -> error::Result<HashMap<String, ServerTimestamp>> {
+            match &self.info_collections {
+                Ok(collections) => Ok(collections.clone()),
+                Err(_) => Err(ErrorKind::StorageHttpError {
+                    code: reqwest::StatusCode::InternalServerError,
+                    route: "info/collections".to_string(),
+                }.into()),
+            }
+        }
+
+        fn fetch_meta_global(&self) -> error::Result<BsoRecord<MetaGlobalRecord>> {
+            match &self.meta_global {
+                Ok(global) => Ok(global.clone()),
+                // TODO(lina): Special handling for 404s, we want to ensure we
+                // handle missing keys and other server errors correctly.
+                Err(_) => Err(ErrorKind::StorageHttpError {
+                    code: reqwest::StatusCode::InternalServerError,
+                    route: "meta/global".to_string(),
+                }.into()),
+            }
+        }
+
+        fn put_meta_global(&self, global: &BsoRecord<MetaGlobalRecord>) -> error::Result<()> {
+            Err(ErrorKind::StorageHttpError {
+                code: reqwest::StatusCode::InternalServerError,
+                route: "meta/global".to_string(),
+            }.into())
+        }
+
+        fn fetch_crypto_keys(&self) -> error::Result<BsoRecord<EncryptedPayload>> {
+            match &self.crypto_keys {
+                Ok(keys) => Ok(keys.clone()),
+                // TODO(lina): Same as above, for 404s.
+                Err(_) => Err(ErrorKind::StorageHttpError {
+                    code: reqwest::StatusCode::InternalServerError,
+                    route: "crypto/keys".to_string(),
+                }.into()),
+            }
+        }
+
+        fn put_crypto_keys(&self, keys: &EncryptedBso) -> error::Result<()> {
+            Err(ErrorKind::StorageHttpError {
+                code: reqwest::StatusCode::InternalServerError,
+                route: "crypto/keys".to_string(),
+            }.into())
+        }
+
+        fn wipe_all(&self) -> error::Result<()> {
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn test_state_machine() {
+        let root_key = KeyBundle::new_random().unwrap();
+        let keys = CollectionKeys {
+            timestamp: 123.4.into(),
+            default: KeyBundle::new_random().unwrap(),
+            collections: HashMap::new(),
+        };
+        let client = InMemoryClient {
+            info_configuration: Ok(InfoConfiguration::default()),
+            info_collections: Ok(vec![("meta", 123.456), ("crypto", 145.0)]
+                .iter()
```

Comment:
`.iter().cloned()` is `into_iter()`, isn't it?

Reply:
No, it's `iter().map(|x| *x)`